rust/kernel/sync/arc.rs

```diff
@@ -7,178 +7,216 @@
 //! underlying object when it reaches zero. It is also safe to use concurrently from multiple
 //! threads.
 //!
-//! It is different from the standard library's [`Arc`] in two ways: it does not support weak
-//! references, which allows it to be smaller -- a single pointer-sized integer; it allows users to
-//! safely increment the reference count from a single reference to the underlying object.
+//! It is different from the standard library's [`Arc`] in a few ways:
+//! 1. It is backed by the kernel's `refcount_t` type.
+//! 2. It does not support weak references, which allows it to be half the size.
+//! 3. It saturates the reference count instead of aborting when it goes over a threshold.
+//! 4. It does not provide a `get_mut` method, so the ref counted object is pinned.
 //!
 //! [`Arc`]: https://doc.rust-lang.org/std/sync/struct.Arc.html
 
-use crate::Result;
+use crate::{bindings, Result};
 use alloc::boxed::Box;
 use core::{
-    mem::ManuallyDrop,
-    ops::Deref,
-    ptr::NonNull,
-    sync::atomic::{fence, AtomicUsize, Ordering},
+    cell::UnsafeCell, convert::AsRef, mem::ManuallyDrop, ops::Deref, pin::Pin, ptr::NonNull,
 };
 
+extern "C" {
+    fn rust_helper_refcount_new() -> bindings::refcount_t;
+    fn rust_helper_refcount_inc(r: *mut bindings::refcount_t);
+    fn rust_helper_refcount_dec_and_test(r: *mut bindings::refcount_t) -> bool;
+}
+
 /// A reference-counted pointer to an instance of `T`.
 ///
 /// The reference count is incremented when new instances of [`Ref`] are created, and decremented
 /// when they are dropped. When the count reaches zero, the underlying `T` is also dropped.
 ///
 /// # Invariants
 ///
-/// The value stored in [`RefCounted::get_count`] corresponds to the number of instances of [`Ref`]
-/// that point to that instance of `T`.
-pub struct Ref<T: RefCounted + ?Sized> {
-    ptr: NonNull<T>,
+/// The reference count on an instance of [`Ref`] is always non-zero.
+/// The object pointed to by [`Ref`] is always pinned.
+pub struct Ref<T: ?Sized> {
+    ptr: NonNull<RefInner<T>>,
```

Comment:
IMO `NonNull` is correct here. `Ref` is conceptually `&T` so it should be covariant. `*mut T` is invariant and is thus not correct. `NonNull` also ensures that `Option<Ref<T>>` can be stored in a single usize.

Reply:
Thanks for the advice Gary! I posted the covariance doc quote, because I don't really understand what it means. Maybe you could provide an example that would not work (or be unsound) if we used `*mut T` instead of `NonNull<T>` ?

And what about the `PhantomData` remark in the docs, do you think we need any of that here?

Reply:
A pretty synthetic example:
```rust
use std::ptr::NonNull;

struct Ref<T: ?Sized> {
    x: NonNull<T>,
    // x: *mut T,
}

fn take<'a>(r: Ref<&'a u32>, y: &'a u32) {}

fn give() -> Ref<&'static u32> { todo!() }

fn test() {
    let y = 1;
    take(give(), &y);
}
```
If you change x to `*mut T` the code won't compile.

With that said, we still need additional `PhantomData` field here, but it's not about variance, but rather drop check: https://doc.rust-lang.org/nightly/std/marker/struct.PhantomData.html#ownership-and-the-drop-check. As `Ref` can potentially cause `T` to be dropped, an `PhantomData` marker is still needed.

Reply:
Thanks Gary, that's awesome! My naive view of covariance is this:
```rust
struct Dog;
struct Cat;

trait Animal {
    fn sound(&self);
}

impl Animal for Dog {
    fn sound(&self) {}
}

impl Animal for Cat {
    fn sound(&self) {}
}

let c = Arc::try_new(Cat)?;
let d = Arc::try_new(Dog)?;
let a: Arc<dyn Animal> = d; // this works ok, covariant
```
But for `Ref<T>` (our `Arc` replacement), that doesn't appear to work, what am I missing?
```rust
let c = Ref::try_new(Cat)?;
let d = Ref::try_new(Dog)?;
let a: Ref<dyn Animal> = c;
```
```
error[E0308]: mismatched types
   --> drivers/android/process.rs:312:34
    |
312 |         let a: Ref<dyn Animal> = c;
    |                ---------------   ^ expected trait object `dyn Animal`, found struct `Cat`
```

Reply:
This piece is missing. Gary helped me with this on zulip, here's a link the solution: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=b1e7c2183fd2ee4f1970b19864d6ed6b

Since it requires 3 _additional_ unstable features and we don't need it immediately (we will need it when I convert the usage of lists of `dyn DeliverToRead` in binder), I thought it would make sense to push it out to its own PR.